crates/ntx-builder/src/builder/server/state.rs

```diff
@@ -0,0 +1,123 @@
+use std::collections::{BTreeMap, BTreeSet, VecDeque};
+
+use miden_node_utils::network_note::NetworkNote;
+use miden_objects::{
+    note::{NoteId, NoteTag, Nullifier},
+    transaction::TransactionId,
+};
+
+/// Max number of notes taken for executing a single network transaction
+const MAX_BATCH: usize = 50;
+
+/// Maintains state of the network notes for the transaction builder.
+///
+/// Note tags are mainly kept in a [`VecDeque`], but also indexed by nullifiers and note tags in
+/// order to enable simple lookups when discarding by nullifiers and deciding which notes to
+/// execute against a unique account, respectively.
+/// Additionally, a list of inflight notes is kept to track their lifecycle.
```

Comment:
I think this description needs to be updated (e.g., we don't index note tags by nullifier).